File: lib/services/bird_repository.dart
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart';
import 'package:sqflite/sqflite.dart' as sql;
import '../models/models.dart';
import 'database_service.dart';

class BirdRepository {
  final DatabaseService _dbService = DatabaseService.instance;

  // Projects
  Future<List<Project>> getAllProjects() async {
    final db = await _dbService.database;
    final maps = await db.query('projects', orderBy: 'updatedAt DESC');
    return maps.map((m) => Project.fromMap(m)).toList();
  }

  Future<Project?> getProjectById(int id) async {
    final db = await _dbService.database;
    final maps = await db.query('projects', where: 'id = ?', whereArgs: [id]);
    if (maps.isNotEmpty) return Project.fromMap(maps.first);
    return null;
  }

  Future<int> saveProject(Project project) async {
    final db = await _dbService.database;
    return await db.insert('projects', project.toMap());
  }

  Future<int> updateProject(Project project) async {
    final db = await _dbService.database;
    return await db.update('projects', project.toMap(),
        where: 'id = ?', whereArgs: [project.id]);
  }

  Future<int> deleteProject(int id) async {
    final db = await _dbService.database;
    return await db.delete('projects', where: 'id = ?', whereArgs: [id]);
  }

  // Sites
  Future<List<Site>> getSitesForProject(int projectId) async {
    final db = await _dbService.database;
    final maps = await db.query('sites',
        where: 'projectId = ?', whereArgs: [projectId], orderBy: 'id DESC');
    return maps.map((m) => Site.fromMap(m)).toList();
  }

  Future<int> saveSite(Site site) async {
    final db = await _dbService.database;
    return await db.insert('sites', site.toMap());
  }

  Future<int> deleteSite(int id) async {
    final db = await _dbService.database;
    return await db.delete('sites', where: 'id = ?', whereArgs: [id]);
  }

  // Samplings
  Future<List<Sampling>> getSamplingsForSite(int siteId) async {
    final db = await _dbService.database;
    final maps = await db.query('samplings',
        where: 'siteId = ?', whereArgs: [siteId], orderBy: 'date DESC');
    return maps.map((m) => Sampling.fromMap(m)).toList();
  }

  Future<int> saveSampling(Sampling sampling) async {
    final db = await _dbService.database;
    return await db.insert('samplings', sampling.toMap());
  }

  Future<int> updateSampling(Sampling sampling) async {
    final db = await _dbService.database;
    return await db.update('samplings', sampling.toMap(),
        where: 'id = ?', whereArgs: [sampling.id]);
  }

  Future<int> deleteSampling(int id) async {
    final db = await _dbService.database;
    return await db.delete('samplings', where: 'id = ?', whereArgs: [id]);
  }

  // Observations
  Future<List<Observation>> getObservationsForSampling(int samplingId) async {
    final db = await _dbService.database;
    final maps = await db.query('observations',
        where: 'samplingId = ?', whereArgs: [samplingId], orderBy: 'timestamp DESC');
    return maps.map((m) => Observation.fromMap(m)).toList();
  }

  Future<List<Observation>> getAllObservations() async {
    final db = await _dbService.database;
    final maps = await db.query('observations', orderBy: 'timestamp DESC');
    return maps.map((m) => Observation.fromMap(m)).toList();
  }

  Future<int> saveObservation(Observation observation) async {
    final db = await _dbService.database;
    return await db.insert('observations', observation.toMap());
  }

  Future<int> deleteObservation(int id) async {
    final db = await _dbService.database;
    return await db.delete('observations', where: 'id = ?', whereArgs: [id]);
  }

  // Sessions
  Future<List<SamplingSession>> getAllSessions() async {
    final db = await _dbService.database;
    final maps = await db.query('sampling_sessions', orderBy: 'createdAt DESC');
    return maps.map((m) => SamplingSession.fromMap(m)).toList();
  }

  Future<int> saveSession(SamplingSession session) async {
    final db = await _dbService.database;
    if (session.id != null && session.id! > 0) {
      await db.update('sampling_sessions', session.toMap(),
          where: 'id = ?', whereArgs: [session.id]);
      return session.id!;
    }
    return await db.insert('sampling_sessions', session.toMap());
  }

  Future<int> deleteSession(int id) async {
    final db = await _dbService.database;
    return await db.delete('sampling_sessions', where: 'id = ?', whereArgs: [id]);
  }

  // Sync Logs
  Future<List<SyncLog>> getAllSyncLogs() async {
    final db = await _dbService.database;
    final maps = await db.query('sync_logs', orderBy: 'timestamp DESC');
    return maps.map((m) => SyncLog.fromMap(m)).toList();
  }

  Future<int> saveSyncLog(SyncLog log) async {
    final db = await _dbService.database;
    return await db.insert('sync_logs', log.toMap());
  }

  // Articles
  Future<List<BirdArticle>> getAllArticles() async {
    final db = await _dbService.database;
    final maps = await db.query('bird_articles', orderBy: 'pubDate DESC');
    return maps.map((m) => BirdArticle.fromMap(m)).toList();
  }

  Future<void> saveArticles(List<BirdArticle> articles) async {
    final db = await _dbService.database;
    final batch = db.batch();
    for (var article in articles) {
      batch.insert('bird_articles', article.toMap(),
          conflictAlgorithm: sql.ConflictAlgorithm.replace);
    }
    await batch.commit(noResult: true);
  }

  Future<bool> fetchAndCacheAllAboutBirdsArticles() async {
    try {
      final response = await http.get(Uri.parse('https://www.allaboutbirds.org/news/feed/'));
      if (response.statusCode == 200) {
        final document = XmlDocument.parse(response.body);
        final items = document.findAllElements('item');
        final articles = <BirdArticle>[];

        for (var item in items) {
          final title = item.findElements('title').firstOrNull?.innerText ?? 'Sin Título';
          final link = item.findElements('link').firstOrNull?.innerText ?? '';
          final pubDate = item.findElements('pubDate').firstOrNull?.innerText ?? '';
          final description = item.findElements('description').firstOrNull?.innerText ?? '';
          final creator = item.findElements('dc:creator').firstOrNull?.innerText ?? 'Cornell Lab';
          final category = item.findElements('category').firstOrNull?.innerText ?? 'Mundo Aves';

          final cleanDesc = description
              .replaceAll('<![CDATA[', '')
              .replaceAll(']]>', '')
              .replaceAll(RegExp(r'<[^>]*>|&[^;]+;'), ' ')
              .trim();

          articles.add(BirdArticle(
            link: link,
            title: title.replaceAll('<![CDATA[', '').replaceAll(']]>', '').trim(),
            description: cleanDesc.length > 200 ? '${cleanDesc.substring(0, 197)}...' : cleanDesc,
            pubDate: pubDate,
            creator: creator,
            category: category,
          ));
        }

        if (articles.isNotEmpty) {
          await saveArticles(articles);
          return true;
        }
      }
    } catch (e) {
      print('Error fetching articles: $e');
    }
    return false;
  }

  // Templates
  Future<List<FormTemplate>> getAllTemplates() async {
    final db = await _dbService.database;
    final maps = await db.query('form_templates');
    return maps.map((m) => FormTemplate.fromMap(m)).toList();
  }

  Future<void> prePopulateTemplates() async {
    final templates = await getAllTemplates();
    if (templates.isEmpty) {
      final db = await _dbService.database;
      await db.insert('form_templates', FormTemplate(
        name: "Monitoreo Estándar de Avifauna",
        fieldsJson: '[{"key": "cloud_cover", "label": "Cobertura de Nubes (%)", "type": "NUMBER", "required": true}, {"key": "canopy_status", "label": "Estado del Dosel", "type": "DROPDOWN", "options": "Denso, Medio, Abierto, Ninguuno", "required": false}, {"key": "wind_speed", "label": "Velocidad del Viento (km/h)", "type": "NUMBER", "required": false}, {"key": "noise_level", "label": "Nivel de Ruido Ambiental", "type": "DROPDOWN", "options": "Bajo, Moderado, Alto", "required": true}]'
      ).toMap());
      await db.insert('form_templates', FormTemplate(
        name: "Ficha de Hábitat y Vegetación",
        fieldsJson: '[{"key": "canopy_height", "label": "Altura Promedio de Dosel (m)", "type": "NUMBER", "required": true}, {"key": "understory_density", "label": "Densidad de Sotobosque", "type": "DROPDOWN", "options": "Alta, Media, Baja, Nula", "required": true}, {"key": "human_impact", "label": "Evidencia de Impacto Humano", "type": "DROPDOWN", "options": "Grave, Moderado, Leve, Ninguno", "required": true}]'
      ).toMap());
    }
  }
}

File: app/src/main/java/com/example/data/BirdRepository.kt
package com.example.data

import android.content.Context
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.firstOrNull

class BirdRepository(private val db: AppDatabase) {

    // Sessions
    val allSessions: Flow<List<SamplingSession>> = db.samplingSessionDao().getAllSessionsFlow()
    suspend fun getSessionsSync(): List<SamplingSession> = db.samplingSessionDao().getAllSessions()
    suspend fun getSessionById(id: Int): SamplingSession? = db.samplingSessionDao().getSessionById(id)
    suspend fun saveSession(session: SamplingSession): Long = db.samplingSessionDao().insertSession(session)
    suspend fun deleteSessionById(id: Int) = db.samplingSessionDao().deleteSessionById(id)

    // Projects
    val allProjects: Flow<List<Project>> = db.projectDao().getAllProjects()
    
    fun getProject(id: Int): Flow<Project?> = db.projectDao().getProjectById(id)
    
    suspend fun getProjectSync(id: Int): Project? = db.projectDao().getProjectByIdSync(id)

    suspend fun saveProject(project: Project): Long = db.projectDao().insertProject(project)

    suspend fun updateProject(project: Project) = db.projectDao().updateProject(project)

    suspend fun deleteProject(project: Project) = db.projectDao().deleteProject(project)

    // Sites
    fun getSitesForProject(projectId: Int): Flow<List<Site>> = db.siteDao().getSitesForProject(projectId)

    fun getSiteById(id: Int): Flow<Site?> = db.siteDao().getSiteById(id)
    
    suspend fun getSitesForProjectSync(projectId: Int): List<Site> = db.siteDao().getSitesForProjectSync(projectId)

    suspend fun getSiteByIdSync(id: Int): Site? = db.siteDao().getSiteByIdSync(id)

    suspend fun saveSite(site: Site): Long = db.siteDao().insertSite(site)

    suspend fun deleteSite(site: Site) = db.siteDao().deleteSite(site)

    // Samplings
    fun getSamplingsForSite(siteId: Int): Flow<List<Sampling>> = db.samplingDao().getSamplingsForSite(siteId)

    fun getSamplingById(id: Int): Flow<Sampling?> = db.samplingDao().getSamplingById(id)

    suspend fun getSamplingsForSiteSync(siteId: Int): List<Sampling> = db.samplingDao().getSamplingsForSiteSync(siteId)

    suspend fun getSamplingByIdSync(id: Int): Sampling? = db.samplingDao().getSamplingByIdSync(id)

    suspend fun saveSampling(sampling: Sampling): Long = db.samplingDao().insertSampling(sampling)

    suspend fun updateSampling(sampling: Sampling) = db.samplingDao().updateSampling(sampling)

    suspend fun deleteSampling(sampling: Sampling) = db.samplingDao().deleteSampling(sampling)

    // Observations
    fun getObservationsForSampling(samplingId: Int): Flow<List<Observation>> = db.observationDao().getObservationsForSampling(samplingId)

    fun getObservationById(id: Int): Flow<Observation?> = db.observationDao().getObservationById(id)

    val allObservations: Flow<List<Observation>> = db.observationDao().getAllObservations()

    suspend fun getObservationsForSamplingSync(samplingId: Int): List<Observation> = db.observationDao().getObservationsForSamplingSync(samplingId)

    suspend fun getAllObservationsSync(): List<Observation> = db.observationDao().getAllObservationsSync()

    suspend fun saveObservation(observation: Observation): Long = db.observationDao().insertObservation(observation)

    suspend fun deleteObservation(observation: Observation) = db.observationDao().deleteObservation(observation)

    // Templates
    val allTemplates: Flow<List<FormTemplate>> = db.formTemplateDao().getTemplates()

    suspend fun saveTemplate(template: FormTemplate): Long = db.formTemplateDao().insertTemplate(template)

    suspend fun deleteTemplate(template: FormTemplate) = db.formTemplateDao().deleteTemplate(template)

    // Sync logs
    val allSyncLogs: Flow<List<SyncLog>> = db.syncLogDao().getSyncLogs()

    suspend fun insertSyncLog(log: SyncLog) = db.syncLogDao().insertSyncLog(log)

    // Articles & News from All About Birds
    val allArticles: Flow<List<BirdArticle>> = db.birdArticleDao().getAllArticles()

    suspend fun saveArticles(articles: List<BirdArticle>) = db.birdArticleDao().insertArticles(articles)

    suspend fun fetchAndCacheAllAboutBirdsArticles(): Boolean {
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            try {
                val client = okhttp3.OkHttpClient.Builder()
                    .connectTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
                    .readTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
                    .build()
                val request = okhttp3.Request.Builder()
                    .url("https://www.allaboutbirds.org/news/feed/")
                    .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) Android BirdWatch/1.0")
                    .build()
                val response = client.newCall(request).execute()
                if (!response.isSuccessful) return@withContext false
                val body = response.body?.string() ?: return@withContext false
                
                val articles = mutableListOf<BirdArticle>()
                val items = body.split("<item>")
                for (i in 1 until items.size) {
                    val itemContent = items[i].split("</item>")[0]
                    
                    val title = extractTagContent(itemContent, "title") ?: "Sin Título"
                    val link = extractTagContent(itemContent, "link") ?: "https://www.allaboutbirds.org/news/articles_${System.currentTimeMillis()}_$i"
                    val pubDate = extractTagContent(itemContent, "pubDate") ?: ""
                    val description = extractTagContent(itemContent, "description") ?: ""
                    val creator = extractTagContent(itemContent, "dc:creator") ?: "Cornell Lab"
                    val category = extractTagContent(itemContent, "category") ?: "Mundo Aves"
                    
                    val cleanDesc = description
                        .replace("<![CDATA[", "")
                        .replace("]]>", "")
                        .replace(Regex("<.*?>"), "") // Strip HTML tag syntax
                        .trim()
                    
                    val cleanTitle = title
                        .replace("<![CDATA[", "")
                        .replace("]]>", "")
                        .trim()

                    articles.add(
                        BirdArticle(
                            link = link,
                            title = cleanTitle,
                            description = if (cleanDesc.length > 200) cleanDesc.take(197) + "..." else cleanDesc,
                            pubDate = pubDate,
                            creator = creator,
                            category = category
                        )
                    )
                }

                if (articles.isNotEmpty()) {
                    db.birdArticleDao().insertArticles(articles)
                    true
                } else {
                    false
                }
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
    }

    private fun extractTagContent(xml: String, tagName: String): String? {
        val openTag = "<$tagName>"
        val closeTag = "</$tagName>"
        val startIndex = xml.indexOf(openTag)
        if (startIndex == -1) {
            val tagRegex = Regex("<$tagName>(.*?)</$tagName>", RegexOption.DOT_MATCHES_ALL)
            val match = tagRegex.find(xml)
            return match?.groupValues?.get(1)
        }
        val endIndex = xml.indexOf(closeTag, startIndex + openTag.length)
        if (endIndex == -1) return null
        return xml.substring(startIndex + openTag.length, endIndex)
    }

    // Check & Prepopulate templates if empty
    suspend fun prePopulateTemplates() {
        val current = db.formTemplateDao().getTemplatesSync()
        if (current.isEmpty()) {
            val template1 = FormTemplate(
                name = "Monitoreo Estándar de Avifauna",
                fieldsJson = """
                    [
                      {"key": "cloud_cover", "label": "Cobertura de Nubes (%)", "type": "NUMBER", "required": true},
                      {"key": "canopy_status", "label": "Estado del Dosel", "type": "DROPDOWN", "options": "Denso, Medio, Abierto, Ninguuno", "required": false},
                      {"key": "wind_speed", "label": "Velocidad del Viento (km/h)", "type": "NUMBER", "required": false},
                      {"key": "noise_level", "label": "Nivel de Ruido Ambiental", "type": "DROPDOWN", "options": "Bajo, Moderado, Alto", "required": true}
                    ]
                """.trimIndent()
            )
            val template2 = FormTemplate(
                name = "Ficha de Hábitat y Vegetación",
                fieldsJson = """
                    [
                      {"key": "canopy_height", "label": "Altura Promedio de Dosel (m)", "type": "NUMBER", "required": true},
                      {"key": "understory_density", "label": "Densidad de Sotobosque", "type": "DROPDOWN", "options": "Alta, Media, Baja, Nula", "required": true},
                      {"key": "human_impact", "label": "Evidencia de Impacto Humano", "type": "DROPDOWN", "options": "Grave, Moderado, Leve, Ninguno", "required": true}
                    ]
                """.trimIndent()
            )
            db.formTemplateDao().insertTemplate(template1)
            db.formTemplateDao().insertTemplate(template2)
        }
    }
}
